contracts/starknet/lib/timestamp.cairo

```diff
@@ -0,0 +1,37 @@
+%lang starknet
+
+from starkware.cairo.common.cairo_builtins import HashBuiltin
+
+@contract_interface
+namespace IL1HeadersStore:
+    func get_latest_l1_block() -> (number : felt):
+    end
+end
+
+@storage_var
+func l1_headers_store() -> (res : felt):
+end
+
+@storage_var
+func timestamp_to_eth_block_number(timestamp : felt) -> (number : felt):
+end
+
+func get_eth_block_number{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+    timestamp : felt
+) -> (number : felt):
+    let (number) = timestamp_to_eth_block_number.read(timestamp)
+    if number != 0:
+        # The timestamp has already be queried in fossil and stored. Therefore we can just return the stored value
+        # This branch will be taken whenever a vote is cast as the mapping value would be set at proposal creation.
+        return (number)
+    else:
+        # The timestamp has not yet been queried in fossil. Therefore we must query Fossil for the latest eth block
```

Comment:
This means a proposer can't create a proposal with a snapshot that was BEFORE the creation of the proposal right? Like if I create a proposal on second of Jan and I want the snapshot to be taken on the 1st of Jan (one day before), that's not possible, correct?